test/hotspot/gtest/gc/g1/test_g1CardSet.cpp

```diff
@@ -0,0 +1,480 @@
+
+    check_iteration(&card_set, card_set.occupied());
+
+    card_set.clear();
+    ASSERT_TRUE(card_set.occupied() == 0);
+  }
+  { // Test coarsening to full
+    G1CardSet card_set(&config, &mm);
+
+    uint count = 0;
+    uint i = 10;
+    uint bitmap_threshold = config.cards_in_howl_bitmap_threshold();
+    for (; i <  bitmap_threshold + 10; i++) {
+      G1AddCardResult res = card_set.add_card(99, i);
+      ASSERT_TRUE(res == Added);
+      count++;
+      ASSERT_TRUE(count == card_set.occupied());
+    }
+
+    G1AddCardResult res = card_set.add_card(99, config.num_cards_in_howl_bitmap() - 1);
+    // Adding above card should have coarsened Bitmap -> Full.
+    ASSERT_TRUE(res == Added);
+    ASSERT_TRUE(config.num_cards_in_howl_bitmap() == card_set.occupied());
+
+    res = card_set.add_card(99, config.num_cards_in_howl_bitmap() - 2);
+    ASSERT_TRUE(res == Found);
+
+    uint threshold = config.cards_in_howl_threshold();
+    uint adjusted_threshold = config.cards_in_howl_bitmap_threshold() * config.num_buckets_in_howl();
+    i = config.num_cards_in_howl_bitmap();
+    count = i;
+    for (; i <  threshold; i++) {
+      G1AddCardResult res = card_set.add_card(99, i);
+      ASSERT_TRUE(res == Added);
+      count++;
+      ASSERT_TRUE(count == card_set.occupied());
+    }
+
+    res = card_set.add_card(99, CardsPerRegion - 1);
+    // Adding above card should have coarsened Howl -> Full.
+    ASSERT_TRUE(res == Added);
+    ASSERT_TRUE(CardsPerRegion == card_set.occupied());
+
+    check_iteration(&card_set, card_set.occupied());
+
+    res = card_set.add_card(99, CardsPerRegion - 2);
+    ASSERT_TRUE(res == Found);
+
+    G1CountCardsClosure count_cards;
+    card_set.iterate_cards(count_cards);
+    ASSERT_TRUE(count_cards._num_cards == config.max_cards_in_region());
+
+    card_set.clear();
+    ASSERT_TRUE(card_set.occupied() == 0);
+  }
+}
+
+class G1CardSetMtTestTask : public AbstractGangTask {
+  G1CardSet* _card_set;
+
+  size_t _added;
+  size_t _found;
+
+public:
+  G1CardSetMtTestTask(G1CardSet* card_set) :
+    AbstractGangTask(""),
+    _card_set(card_set),
+    _added(0),
+    _found(0) { }
+
+  void work(uint worker_id) {
+    uint seed = worker_id;
+    size_t added = 0;
+    size_t found = 0;
+
+    for (uint i = 0; i < 100000; i++) {
+      uint region = G1CardSetTest::next_random(seed, 1000);
+      uint card = G1CardSetTest::next_random(seed, 10000);
+
+      G1AddCardResult res = _card_set->add_card(region, card);
+
+      ASSERT_TRUE(res == Added || res == Found);
+      if (res == Added) {
+        added++;
+      } else if (res == Found) {
+        found++;
+      }
+    }
+    Atomic::add(&_added, added);
+    Atomic::add(&_found, found);
+  }
+
+  size_t added() const { return _added; }
+  size_t found() const { return _found; }
+};
+
+void G1CardSetTest::cardset_mt_test() {
+  const uint CardsPerRegion = 16384;
+  const double FullCardSetThreshold = 1.0;
+  const uint BitmapCoarsenThreshold = 1.0;
```

Comment:
Since that test randomly adds cards, it is very hard to calculate the expected number of cards for verification if we do not know where the coarsening exactly happens. This is very complicated to test, and other tests already test the coarsening, although not in an MT context, so I would like to not spend the time for either a brittle or useless test.